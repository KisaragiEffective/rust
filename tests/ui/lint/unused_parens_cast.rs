// check-pass

#![warn(unused_parens)]

struct Foo(f32);

impl Foo {
    pub fn f32(self) -> f32 {
        64.0f32
    }
}

fn bar() -> f32 {
    3.0f32
}

mod inner {
    pub mod yet_inner {
        pub mod most_inner {
            pub static VERY_LONG_PATH: f32 = 99.0f32;
        }
    }
}

fn basic_test() {
    // should fire
    let one = 1.0f32;
    let _ = (one) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    let _ = (inner::yet_inner::most_inner::VERY_LONG_PATH) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    let _ = (Foo(1.0f32).0) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    let _ = (Foo(1.0f32).f32()) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    let _ = (bar()) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    let baz = [4.0f32];
    let _ = (baz[0]) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    // following is technically unnecessary, but is allowed because it may confusing.
    let _ = (-1.0f32) as f64;
    let x = Box::new(-1.0f32);
    let _ = (*x) as f64;
    //~^ WARN unnecessary parentheses around cast expression
    // cast is left-assoc
    let _ = (true as u8) as u16;
    //~^ WARN unnecessary parentheses around cast expression
    // should not fire
    let _ = (1.0f32 * 2.0f32) as f64;
    let _ = (1.0f32 / 2.0f32) as f64;
    let _ = (1.0f32 % 2.0f32) as f64;
    let _ = (1.0f32 + 2.0f32) as f64;
    let _ = (1.0f32 - 2.0f32) as f64;
    let _ = (42 << 1) as i64;
    let _ = (42 >> 1) as i64;
    let _ = (42 & 0x1F) as f64;
    let _ = (42 ^ 0x1F) as f64;
    let _ = (42 | 0x1F) as f64;
    let _ = (1.0f32 == 2.0f32) as u8;
    let _ = (1.0f32 != 2.0f32) as u8;
    let _ = (1.0f32 < 2.0f32) as u8;
    let _ = (1.0f32 > 2.0f32) as u8;
    let _ = (1.0f32 <= 2.0f32) as u8;
    let _ = (1.0f32 >= 2.0f32) as u8;
    let _ = (true && false) as u8;
    let _ = (true || false) as u8;
    // skipped range: `as`-cast does not allow non-primitive cast
    // also skipped compound operator
}

fn issue_88519() {
    let _ = ({ 1 }) as i64;
    let _ = (match 0 { x => x }) as i64;
    let _ = (if true { 16 } else { 42 }) as i64;
}

fn issue_51185() -> impl Into<for<'a> fn(&'a ())> {
    // removing parens will change semantics, and make compile does not pass
    (|_| {}) as for<'a> fn(&'a ())
}

fn issue_clippy_10557() {
    let x = 0f32;
    let y = 0f32;
    let width = 100f32;
    let height = 100f32;

    new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
    //~^ WARN unnecessary parentheses around cast expression
    //~^^ WARN unnecessary parentheses around cast expression
    //~^^^ WARN unnecessary parentheses around cast expression
    //~^^^^ WARN unnecessary parentheses around cast expression
}

fn new_rect(x: f64, y: f64, width: f64, height: f64) {

}

fn main() {
}
