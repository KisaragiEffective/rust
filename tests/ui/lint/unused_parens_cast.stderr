warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:28:13
   |
LL |     let _ = (one) as f64;
   |             ^   ^
   |
note: the lint level is defined here
  --> $DIR/unused_parens_cast.rs:3:9
   |
LL | #![warn(unused_parens)]
   |         ^^^^^^^^^^^^^
help: remove these parentheses
   |
LL -     let _ = (one) as f64;
LL +     let _ = one as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:30:13
   |
LL |     let _ = (inner::yet_inner::most_inner::VERY_LONG_PATH) as f64;
   |             ^                                            ^
   |
help: remove these parentheses
   |
LL -     let _ = (inner::yet_inner::most_inner::VERY_LONG_PATH) as f64;
LL +     let _ = inner::yet_inner::most_inner::VERY_LONG_PATH as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:32:13
   |
LL |     let _ = (Foo(1.0f32).0) as f64;
   |             ^             ^
   |
help: remove these parentheses
   |
LL -     let _ = (Foo(1.0f32).0) as f64;
LL +     let _ = Foo(1.0f32).0 as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:34:13
   |
LL |     let _ = (Foo(1.0f32).f32()) as f64;
   |             ^                 ^
   |
help: remove these parentheses
   |
LL -     let _ = (Foo(1.0f32).f32()) as f64;
LL +     let _ = Foo(1.0f32).f32() as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:36:13
   |
LL |     let _ = (bar()) as f64;
   |             ^     ^
   |
help: remove these parentheses
   |
LL -     let _ = (bar()) as f64;
LL +     let _ = bar() as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:39:13
   |
LL |     let _ = (baz[0]) as f64;
   |             ^      ^
   |
help: remove these parentheses
   |
LL -     let _ = (baz[0]) as f64;
LL +     let _ = baz[0] as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:41:13
   |
LL |     let _ = (-1.0f32) as f64;
   |             ^       ^
   |
help: remove these parentheses
   |
LL -     let _ = (-1.0f32) as f64;
LL +     let _ = -1.0f32 as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:44:13
   |
LL |     let _ = (*x) as f64;
   |             ^  ^
   |
help: remove these parentheses
   |
LL -     let _ = (*x) as f64;
LL +     let _ = *x as f64;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:47:13
   |
LL |     let _ = (true as u8) as u16;
   |             ^          ^
   |
help: remove these parentheses
   |
LL -     let _ = (true as u8) as u16;
LL +     let _ = true as u8 as u16;
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:89:14
   |
LL |     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
   |              ^ ^
   |
help: remove these parentheses
   |
LL -     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
LL +     new_rect(x as f64, (y) as f64, (width) as f64, (height) as f64);
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:89:26
   |
LL |     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
   |                          ^ ^
   |
help: remove these parentheses
   |
LL -     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
LL +     new_rect((x) as f64, y as f64, (width) as f64, (height) as f64);
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:89:38
   |
LL |     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
   |                                      ^     ^
   |
help: remove these parentheses
   |
LL -     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
LL +     new_rect((x) as f64, (y) as f64, width as f64, (height) as f64);
   |

warning: unnecessary parentheses around cast expression
  --> $DIR/unused_parens_cast.rs:89:54
   |
LL |     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
   |                                                      ^      ^
   |
help: remove these parentheses
   |
LL -     new_rect((x) as f64, (y) as f64, (width) as f64, (height) as f64);
LL +     new_rect((x) as f64, (y) as f64, (width) as f64, height as f64);
   |

warning: 13 warnings emitted

